### PR TITLE
Fix #3678 display error on loading resources failed

### LINKED ISSUE #3678: When /play/level/<level>?session=<id> doesn't find a session, no error shows
Try this: http://codecombat.com/play/level/sarven-outpost?session=573d6898746e452100edc7ce

There's a 404 because that session doesn't exist, but the loading screen is just blank; no error feedback.

Suggested solution:

1. Have LevelLoader emit an error when a resource fails to load
1. Have PlayLevelView listen for that event, and render the error on the screen when that happens.

### PR DESCRIPTION
**Pull request by duybkict**

### CODE REVIEW

File: app/lib/LevelLoader.coffee [hunk 1]
```diff
@@ -322,6 +323,10 @@ module.exports = class LevelLoader extends CocoClass
     @worldNecessities = (r for r in @worldNecessities when r?)
     @onWorldNecessitiesLoaded() if @checkAllWorldNecessitiesRegisteredAndLoaded()
 
+  onWorldNecessityLoadFailed: (resource) ->
+    @trigger('world-necessity-load-failed', resource: resource)
+    return if @headless
```
Comment (sderickson) [inline app/lib/LevelLoader.coffee]: What's this line for?

File: app/views/play/level/PlayLevelView.coffee [hunk 2]
```diff
@@ -195,6 +196,10 @@ module.exports = class PlayLevelView extends RootView
     @controlBar.setBus(@bus)
     @initScriptManager()
 
+  onWorldNecessityLoadFailed: (resource) ->
+    @$el.find('.level-loading-goals, .tip, .load-progress').hide()
+    @$el.find('.could-not-load').show()
```
Comment (sderickson) [inline app/views/play/level/PlayLevelView.coffee]: Elements in a view should only be manipulated by their own view. In this case, the PlayLevelView is manipulating elements in LevelLoadingView. LevelLoadingView should have a function like this (maybe instead called `showCouldNotLoadError`), that PlayLevelView calls when it gets the error from LevelLoader.

**Comment by duybkict** [comment]
How could i pass `levelLoader` or the event `world-necessity-load-failed` from `PlayLevelView` to `LevelLoadingView`?

**Comment by sderickson** [comment]
In PlayLevelView's event handler, have it call a function on LevelLoader. So

```
  onWorldNecessityLoadFailed: (resource) ->
    @loadingView.showCouldNotLoadError()
```

And in LevelLoadingView

```
  showCouldNotLoadError: ->
    @$('.level-loading-goals, .tip, .load-progress').hide()
    @$('.could-not-load').show()
```

Also, just noticed, `@$` is a shorthand you can use there instead of `@$el.find`. It's like the global $ function but scoped to the view's `el`.

**Comment by duybkict** [comment]
hmm, I think I will stick with `@$el.find` as it would produce slightly better performance

**Comment by sderickson** [comment]
Aren't they effectively the same, a find scoped to the view's element?

Either way, looks good to me! However, this branch seems really polluted with other changes that don't make it into the diff, particularly with vi.coffee changes. Can you turn this PR into a single commit? While on your branch, `git rebase master -i` should let you do it, although when I tried to do that it lead to conflicts. It might be better to make a new branch and [apply a patch](https://ariejan.net/2009/10/26/how-to-create-and-apply-a-patch-with-git/) of this branch's difference with master on it.

**Comment by duybkict** [comment]
#3716

**Pull request closed without merging**